jbmc/src/java_bytecode/remove_exceptions.cpp

```diff
@@ -593,24 +593,85 @@ void remove_exceptionst::operator()(goto_programt &goto_program)
 /// removes throws/CATCH-POP/CATCH-PUSH
 void remove_exceptions(
   symbol_table_baset &symbol_table,
-  const class_hierarchyt *class_hierarchy,
   goto_functionst &goto_functions,
-  message_handlert &message_handler,
-  remove_exceptions_typest type)
+  message_handlert &message_handler)
 {
   const namespacet ns(symbol_table);
   std::map<irep_idt, std::set<irep_idt>> exceptions_map;
+
   uncaught_exceptions(goto_functions, ns, exceptions_map);
+
   remove_exceptionst::function_may_throwt function_may_throw =
     [&exceptions_map](const irep_idt &id) {
       return !exceptions_map[id].empty();
     };
+
   remove_exceptionst remove_exceptions(
-    symbol_table,
-    class_hierarchy,
-    function_may_throw,
-    type == remove_exceptions_typest::REMOVE_ADDED_INSTANCEOF,
-    message_handler);
+    symbol_table, nullptr, function_may_throw, false, message_handler);
+
+  remove_exceptions(goto_functions);
+}
+
+/// removes throws/CATCH-POP/CATCH-PUSH from a single GOTO program, replacing
+/// them with explicit exception propagation.
+/// Note this is somewhat less accurate than the whole-goto-model version,
+/// because we can't inspect other functions to determine whether they throw
+/// or not, and therefore must assume they do and always introduce post-call
+/// exception dispatch.
+/// \param goto_program: program to remove exceptions from
+/// \param symbol_table: global symbol table. The `@inflight_exception` global
+///   may be added if not already present. It will not be initialised; that is
+///   the caller's responsibility.
+/// \param message_handler: logging output
+void remove_exceptions(
+  goto_programt &goto_program,
+  symbol_table_baset &symbol_table,
+  message_handlert &message_handler)
+{
+  remove_exceptionst::function_may_throwt any_function_may_throw =
+    [](const irep_idt &) { return true; };
+
+  remove_exceptionst remove_exceptions(
+    symbol_table, nullptr, any_function_may_throw, false, message_handler);
+
+  remove_exceptions(goto_program);
+}
+
+/// removes throws/CATCH-POP/CATCH-PUSH, replacing them with explicit exception
+/// propagation.
+/// \param goto_model: model to remove exceptions from. The
+///   `@inflight_exception` global may be added to its symbol table if not
+///   already present. It will not be initialised; that is the caller's
+///   responsibility.
+/// \param message_handler: logging output
+void remove_exceptions(
+  goto_modelt &goto_model,
+  message_handlert &message_handler)
+{
+  remove_exceptions(
+    goto_model.symbol_table, goto_model.goto_functions, message_handler);
+}
+
+/// removes throws/CATCH-POP/CATCH-PUSH
+void remove_exceptions_and_instanceof(
```

Comment:
Name not really accurate; this should only affect the newly introduced instanceof instructions, not existing ones